core/pta/attestation.c

```diff
@@ -0,0 +1,182 @@
+ */
+static bool is_region_valid(struct vm_region *r)
+{
+	uint32_t skip_flags = VM_FLAG_EPHEMERAL | VM_FLAG_PERMANENT |
+			      VM_FLAG_LDELF;
+
+	return !(r->flags & skip_flags || r->attr & TEE_MATTR_UW);
+}
+
+/*
+ * With this comparison function, we're hashing the smaller regions first.
+ * Regions of equal size are ordered based on their content (memcmp()).
+ * Identical regions can be in any order since they will yield the same hash
+ * anyways.
+ */
+static int cmp_regions(const void *a, const void *b)
+{
+	const struct vm_region *r1 = *(const struct vm_region **)a;
+	const struct vm_region *r2 = *(const struct vm_region **)b;
+
+	if (r1->size < r2->size)
+		return -1;
+
+	if (r1->size > r2->size)
+		return 1;
+
+	return memcmp((void *)r1->va, (void *)r2->va, r1->size);
+}
+
+static TEE_Result hash_regions(struct vm_info *vm_info, uint8_t *hash)
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct vm_region *r = NULL;
+	struct vm_region **regions = NULL;
+	size_t nregions = 0;
+	void *ctx = NULL;
+	size_t i = 0;
+
+	res = crypto_hash_alloc_ctx(&ctx, TEE_ALG_SHA256);
+	if (res)
+		return res;
+
+	res = crypto_hash_init(ctx);
+	if (res)
+		goto out;
+
+	/*
+	 * Make an array of region pointers so we can use qsort() to order it.
+	 */
+
+	TAILQ_FOREACH(r, &vm_info->regions, link)
+		if (is_region_valid(r))
+			nregions++;
+
+	regions = malloc(nregions * sizeof(*regions));
+	if (!regions) {
+		res = TEE_ERROR_OUT_OF_MEMORY;
+		goto out;
+	}
+
+	TAILQ_FOREACH(r, &vm_info->regions, link)
+		if (is_region_valid(r))
+			regions[i++] = r;
+
+	/*
+	 * Sort regions so that they are in a consistent order even when TA ASLR
+	 * is enabled.
+	 */
+	qsort(regions, nregions, sizeof(*regions), cmp_regions);
+
+	/* Hash regions in order */
+	for (i = 0; i < nregions; i++) {
+		r = regions[i];
+		DMSG("va %p size %zu", (void *)r->va, r->size);
+		res = crypto_hash_update(ctx, (uint8_t *)r->va, r->size);
+		if (res)
+			goto out;
+	}
+
+	res = crypto_hash_final(ctx, hash, TEE_SHA256_HASH_SIZE);
+out:
+	free(regions);
+	crypto_hash_free_ctx(ctx);
+	return res;
+}
+
+static TEE_Result hash_ta(uint32_t types, TEE_Param params[TEE_NUM_PARAMS])
+{
+	struct tee_ta_session_head *open_sessions = NULL;
+	struct tee_ta_session *s = NULL;
+	struct ts_session *ts_sess = NULL;
+	uint8_t *uuid_buf = params[0].memref.buffer;
+	size_t uuid_sz = params[0].memref.size;
+	uint32_t id = params[1].value.a;
+	uint8_t *hash = params[2].memref.buffer;
+	size_t hash_sz = params[2].memref.size;
+	TEE_Result ret = TEE_SUCCESS;
+	struct user_ta_ctx *ctx = NULL;
+
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INPUT,
+				     TEE_PARAM_TYPE_VALUE_INPUT,
+				     TEE_PARAM_TYPE_MEMREF_OUTPUT,
+				     TEE_PARAM_TYPE_NONE))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (uuid_sz != sizeof(TEE_UUID))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (!hash || hash_sz != TEE_SHA256_HASH_SIZE)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	nsec_sessions_list_head(&open_sessions);
```

Comment:
It looks like this command is reserved to REE kernel clients.
Session ID is a private information usually not shared with non-secure user land client. Do I get it wrong? 

I wonder also if session IDs related to TA clients should be visible/measurable by REE.

Reply:
> It looks like this command is reserved to REE kernel clients.

No, it is meant for a CA wanting to authenticate the TA it is "talking" to.

> Session ID is a private information usually not shared with non-secure user land client. Do I get it wrong?

The `session_id` member of `TEEC_Session` is definitely public in `tee_client_api.h`. The exact content of `TEEC_Session` is implementation-defined as per the GP spec, so a CA using this feature would definitely not be portable to other TEEs, but we don't care much I suppose.

> I wonder also if session IDs related to TA clients should be visible/measurable by REE.

What do you mean? Measure the CA?

Reply:
A CA should not be allowed to measure other TA sessions than the one it's using. Here, a CA can invoke this PTA providing any session identifier value.

Reply:
> A CA should not be allowed to measure other TA sessions than the one it's using. Here, a CA can invoke this PTA providing any session identifier value.

I agree. I need to add something to prevent that.